design/0003-byop.md

```diff
@@ -0,0 +1,127 @@
+have a prometheus installation paired up with a long term monitroing solutions like Cortex,InfluxDB, or third party
+solutions like Grafana Cloud, etc.
+In those cases prometheus is just a temporary component, that keeps pushing metrics into their long term solution, making users want to re-use their exisitng prometheus.
+
+The following are the list of problems:
+
+- Hard requirement of linkerd-prometheus.
+
+- Configuring the current linkerd-prometheus for remote-writes, etc is hard, because the cm is overriden after upgrades, etc.
+
+- Not enough documentation for bringing your own prometheus case.
+
+Once the RFC is implemented, Apart from the default prometheus enabled installation Users should be
+able to use their exising prometheus instance, to also scrape metrics from the proxy
+
+# Design proposal
+
+[design-proposal]: #design-proposal
+
+## Prometheus as an Add-On
+
+By moving Prometheus as an add-on, We can make it as an optional component that can be used both through Helm and Linkerd CLI.
+The add-on model would well with upgrades, etc as per [#3955](https://github.com/linkerd/linkerd2/pull/3955). This is done by
+moving the prometheus components into a separate sub-chart in the linkerd repo
+and updating the code. The same health checks are also expected, but as a separate
+check Category, which would be transient i.e runs only when it recognises
+that the add-on is enabled.
+
+By default, the prometheus add-on will be **enabled**.
+
+This also provides a single place of configuration for all things related to
+linkerd-prometheus. Though we are making linkerd-prometheus optional but having it is great for Linkerd and external users, for reasons like separation of
+concerns, etc. By allowing the `remote-write`, `alerts`, etc configuration in the
+linkerd-prometheus, we would decrease the need/want for external users to tie their existing prometheus to Linkerd,
+as it would be even easier for them to configure `remote-write`, etc
+configuration from their exisiting prometheus to the linkerd-prometheus right from the start.
+
+The following default prometheus configuration is planned:
+
+```yaml
+prometheus:
+  enabled: true
+  name: linkerd-prometheus
+  url: 
+  image: 
+  global:
+    <global-prom-config> https://prometheus.io/docs/prometheus/latest/configuration/configuration/#configuration-file
+  remote-write:
+    <remote-write-config> https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write
+  alerts:
+    <alert-manager-config> https://prometheus.io/docs/prometheus/latest/configuration/configuration/#alertmanager_config 
+```
+
+All the current default configuration will be moved to `values.yaml` in the above format, allowing users to override this by using Helm or `--addon-config` through CLI.
+
+## Using Exisiting Prometheus
+
+First, Documentation will be provided at `linkerd.io` to guide users on adding
```

Comment:
I'm not sure it is in-scope for this, but it would be awesome if we could make the scrap config update automatically like we've done for the CLI args and help.

Reply:
From what I've spoke to @alpeb, This was definitely considered in-scope for the problem. We could definitely write some CLI tools for supporting, or even automatically updating the prom config with the current scrape-config, but having documentation on what the current scrape-config is, how to add it, etc should be good right?